engine/runtime/src/main/java/org/enso/interpreter/runtime/state/ExecutionEnvironment.java

```diff
@@ -53,49 +47,39 @@ public ExecutionEnvironment withContextDisabled(Atom context) {
   }
 
   private ExecutionEnvironment update(Atom context, boolean value) {
-    assert context.getConstructor().getType()
-        == EnsoContext.get(null).getBuiltins().context().getType();
-    int keyFound = -1;
-    for (int i = 0; i < keys.length; i++) {
-      if (keys[i].equals(context.getConstructor().getName())) {
-        keyFound = i;
-      }
-    }
-    String[] keys1;
-    Boolean[] permissions1;
-    if (keyFound != -1) {
-      keys1 = cloneArray(keys, new String[keys.length]);
-      permissions1 = cloneArray(permissions, new Boolean[keys.length]);
-      permissions1[keyFound] = value;
+    var contextBuiltin = EnsoContext.get(null).getBuiltins().context();
+    assert context.getConstructor().getType() == contextBuiltin.getType();
+    var ctor = context.getConstructor();
+    Permissions newPermissions;
+    if (ctor == contextBuiltin.getInput()) {
+      newPermissions = new Permissions(value, permissions.output, permissions.dataflowStacktrace);
+    } else if (ctor == contextBuiltin.getOutput()) {
+      newPermissions = new Permissions(permissions.input, value, permissions.dataflowStacktrace);
+    } else if (ctor == contextBuiltin.getDataflowStackTrace()) {
+      newPermissions = new Permissions(permissions.input, permissions.output, value);
     } else {
-      keys1 = cloneArray(keys, new String[keys.length + 1]);
-      permissions1 = cloneArray(permissions, new Boolean[keys.length + 1]);
-      keyFound = keys.length;
-      keys1[keyFound] = context.getConstructor().getName();
-      permissions1[keyFound] = value;
+      throw CompilerDirectives.shouldNotReachHere("Unknown context `" + ctor.getName() + "`");
     }
-    return new ExecutionEnvironment(name, keys1, permissions1);
+    return new ExecutionEnvironment(name, newPermissions);
   }
 
-  private <T> T[] cloneArray(T[] fromArray, T[] toArray) {
-    for (int i = 0; i < fromArray.length; i++) {
-      toArray[i] = fromArray[i];
-    }
-    return toArray;
-  }
-
-  public Boolean hasContextEnabled(String context) {
-    int keyFound = -1;
-    for (int i = 0; i < keys.length; i++) {
-      if (keys[i].equals(context)) {
-        keyFound = i;
-      }
-    }
-    if (keyFound != -1) {
-      return permissions[keyFound];
-    } else {
-      return false;
+  /**
+   * Checks if the context is enabled in this execution environment.
+   *
+   * @param runtimeCtx Constructor of {@code Standard.Base.Runtime.Context} builtin type.
+   * @return {@code true} if the context is enabled in this execution environment.
+   */
+  public boolean hasContextEnabled(AtomConstructor runtimeCtx, EnsoContext ensoCtx) {
+    var contextBuiltin = ensoCtx.getBuiltins().context();
```

Comment:
This might better be a node `HasContextEnabledNode` as well.

Reply:
Execution environment handling implemented via nodes in dd25a6f256918a675898e23ca95172ab3d5d463a